File: UI/Feed.cs
using System;
using System.Collections.Generic;
using System.Linq;

using MD.Data;

namespace MD.UI
{
    /// <summary>
    /// Contains functions related to feeds.
    /// </summary>
    public static class Feed
    {
        /// <summary>
        /// Gets an event feed that never fires.
        /// </summary>
        public static EventFeed<T> Null<T>()
        {
            return NullEventFeed<T>.Instance;
        }

        /// <summary>
        /// Maps events in this event feed based on the given mapping function.
        /// </summary>
        public static EventFeed<T> Map<TSource, T>(this EventFeed<TSource> Source, Func<TSource, T> Map)
        {
            throw new NotImplementedException();
        }

        /// <summary>
        /// Maps values in a signal feed based on the given mapping function.
        /// </summary>
        public static SignalFeed<T> Map<TSource, T>(this SignalFeed<TSource> Source, Func<TSource, T> Map)
        {
            throw new NotImplementedException();
        }

        /// <summary>
        /// Filters events in this event feed that don't satisfy the given filter function.
        /// </summary>
        public static EventFeed<T> Filter<T>(this EventFeed<T> Source, Func<T, bool> Filter)
        {
            throw new NotImplementedException();
        }

        /// <summary>
        /// Strips data from this event feed so that it only gives information about when events occur.
        /// </summary>
        public static EventFeed<Void> Strip<T>(this EventFeed<T> Source)
        {
            throw new NotImplementedException();
        }

        /// <summary>
        /// Gets an event feed that fires an event whenever this feed rises (changes from false to true), or returns null if the feed varies continuously.
        /// </summary>
        public static EventFeed<Void> Rising(this SignalFeed<bool> Feed)
        {
            EventFeed<Change<bool>> delta = Feed.Delta;
            if (delta != null)
            {
                return delta.Filter(x => x.New == true).Strip();
            }
            return null;
        }

        /// <summary>
        /// Gets an event feed that fires an event whenever this feed falls (changes from true to false), or returns null if the feed varies continuously.
        /// </summary>
        public static EventFeed<Void> Falling(this SignalFeed<bool> Feed)
        {
            EventFeed<Change<bool>> delta = Feed.Delta;
            if (delta != null)
            {
                return delta.Filter(x => x.New == false).Strip();
            }
            return null;
        }
    }

    /// <summary>
    /// A source of dynamic, time-dependant information of a certain type.
    /// </summary>
    public interface Feed<T>
    {

    }

    /// <summary>
    /// A dynamic, time-dependant value.
    /// </summary>
    public interface SignalFeed<T> : Feed<T>
    {
        /// <summary>
        /// Gets the current value of the feed. The value returned by this property should not be affected by future changes in the feed.
        /// </summary>
        T Current { get; }

        /// <summary>
        /// Gets an event feed that gives an event for each discrete change in this feed. This will return null for signal feeds that
        /// change continously, with no discernible moment of change.
        /// </summary>
        /// <remarks>During an event in this feed, the signal feed will have its new value.</remarks>
        EventFeed<Change<T>> Delta { get; }
    }

    /// <summary>
    /// Contains information about a change in a signal feed.
    /// </summary>
    public struct Change<T>
    {
        public Change(T Old, T New)
        {
            this.Old = Old;
            this.New = New;
        }

        /// <summary>
        /// The old value of the feed.
        /// </summary>
        public T Old;

        /// <summary>
        /// The new value of the feed.
        /// </summary>
        public T New;
    }

    /// <summary>
    /// A source of discrete events.
    /// </summary>
    public interface EventFeed<T> : Feed<T>
    {
        /// <summary>
        /// Registers a callback to be called whenever an event occurs in this feed.
        /// </summary>
        /// <returns>An action that will unregister the callback, or null if the callback will never be called.</returns>
        RetractAction Register(Action<T> Callback);
    }

    /// <summary>
    /// A finite, dynamic, time-dependant set of items of the given type, using the given equality definition.
    /// </summary>
    public interface SetFeed<T, TEquality> : SignalFeed<Set<T, TEquality>>
        where TEquality : Equality<T>
    {

        /// <summary>
        /// Registers a callback to be called for every item currently in the set, and all future items that enter the set. When an item is removed from
        /// a set, the unregister action returned from the callback will be called.
        /// </summary>
        /// <returns>An action that will unregister every item, and will prevent the callback from being 
        /// called for future items that enter the set.</returns>
        RetractAction RegisterMaintain(RegisterItemAction<T> Callback);
    }

    /// <summary>
    /// A function that registers an item in a set using a user-defined method.
    /// </summary>
    /// <returns>An action that will unregister the item, or null if not required.</returns>
    public delegate RetractAction RegisterItemAction<T>(T Item);

    /// <summary>
    /// An event feed that never fires.
    /// </summary>
    public sealed class NullEventFeed<T> : EventFeed<T>
    {
        private NullEventFeed()
        {

        }

        /// <summary>
        /// The only instance of this class.
        /// </summary>
        public static readonly NullEventFeed<T> Instance = new NullEventFeed<T>();

        public RetractAction Register(Action<T> Callback)
        {
            return null;
        }
    }

    /// <summary>
    /// A signal feed that maintains a manually-set value.
    /// </summary>
    public sealed class ControlSignalFeed<T> : SignalFeed<T>
    {
        public ControlSignalFeed(T Initial)
        {
            this._Current = Initial;
        }

        /// <summary>
        /// Gets or sets the current value of this controlled signal feed.
        /// </summary>
        public T Current
        {
            get
            {
                return this._Current;
            }
            set
            {
                T old = this._Current;
                if (this._Delta != null)
                {
                    this._Delta.Fire(new Change<T>(old, value));
                }
                this._Current = value;
            }
        }

        public EventFeed<Change<T>> Delta
        {
            get
            {
                // Do not create a delta feed until requested.
                if (this._Delta == null)
                    this._Delta = new ControlEventFeed<Change<T>>();
                return this._Delta;
            }
        }

        private T _Current;
        private ControlEventFeed<Change<T>> _Delta;
    }

    /// <summary>
    /// A signal feed that gives manually-fired events.
    /// </summary>
    public sealed class ControlEventFeed<T> : EventFeed<T>
    {
        public ControlEventFeed()
        {

        }

        /// <summary>
        /// Causes this event feed to fire an event with the given value.
        /// </summary>
        public void Fire(T Value)
        {
            if (this._Callback != null)
            {
                this._Callback(Value);
            }
        }

        public RetractAction Register(Action<T> Callback)
        {
            this._Callback += Callback;
            return delegate { this._Callback -= Callback; };
        }

        private Action<T> _Callback;
    }
}

File: Source/Feed.fs
namespace MD

open System
open System.Threading
open System.Collections.Generic

/// An interface to a source of discrete events of a certain type.
type EventFeed<'a> =

    /// Registers a callback to be called whenever an event occurs in this
    /// feed. The returned retract action can be used to remove the callback, but
    /// is not guranteed to stop calls to it.
    abstract member Register : ('a -> unit) -> RetractAction

/// Contains information about a change in a value of a certain type.
type Change<'a> = {
    Old : 'a
    New : 'a
    }

/// An interface to a dynamic value of a certain type.
type SignalFeed<'a> =

    /// Gets the current value of this feed. Note that the returned value will not be influenced
    /// by future changes to the signal feed; it acts as an immutable snapshot of the feed
    /// at the current moment.
    abstract member Current : 'a

    /// Gets an event feed that fires an event for every change that occurs in this signal feed, or None 
    /// if not possible (the signal changes continuously or it is derived from another signal with unknown properties).
    /// During an event fired by the returned feed, this signal will use the new value as the current value. Note that 
    /// this feed may be fired even when there is no change, in which case, 
    /// New and Old on the given change will be the same.
    abstract member Delta : EventFeed<Change<'a>> option

/// An interface to a dynamic collection of objects of a certain type.
type CollectionFeed<'a> =

    /// Registers a callback to be called on every current item in the collection, and every
    /// time an item is added. When an item is removed, the retract action returned by the corresponding
    /// item registration action will be called. When the returned retract action of this method is called,
    /// the retract action of every registered item will be called.
    abstract member Register : ('a -> RetractAction) -> RetractAction

// Create type abbreviations
type 'a change = Change<'a>
type 'a signal = SignalFeed<'a>
type 'a event = EventFeed<'a>
type 'a collection = CollectionFeed<'a>

/// An event feed that never fires an event.
type NullEventFeed<'a> private () =
    static let instance = new NullEventFeed<'a> ()

    /// Gets the only instance of this type.
    static member Instance = instance

    interface EventFeed<'a> with
        member this.Register x = null

/// A signal feed that maintains a constant value.
type ConstSignalFeed<'a> (value : 'a) =

    /// Gets the value of this constant signal feed.
    member this.Value = value
    
    interface SignalFeed<'a> with
        member this.Current = value
        member this.Delta = Some (NullEventFeed<Change<'a>>.Instance :> 'a change event)

/// An event feed that allows events to be manually fired.
type ControlEventFeed<'a> () =
    let callbacks = new Registry<'a -> unit> ()

    /// Fires the given event in this feed.
    member this.Fire event = callbacks.Forall (fun a -> a event)

    interface EventFeed<'a> with
        member this.Register callback = callbacks.Add callback

/// A signal feed that maintains a manually set value.
type ControlSignalFeed<'a> (initial : 'a) =
    let mutable value : 'a = initial
    let delta = new ControlEventFeed<'a change> ()

    /// Gets or sets the current value of this feed.
    member this.Current
        with get () = value
        and set x =
            let change = { Old = value; New = x }
            value <- x
            delta.Fire change

    interface SignalFeed<'a> with
        member this.Current = this.Current
        member this.Delta = Some (delta :> 'a change event)

/// A collection feed that allows items to be manually added or removed.
type ControlCollectionFeed<'a> () =
    let items = new LinkedList<'a> ()
    let callbacks = new LinkedList<'a -> RetractAction> ()
    let relations = new LinkedList<LinkedListNode<'a> * LinkedListNode<'a -> RetractAction> * RetractAction> ()

    /// Adds an item to this collection feed and returns a retract handler to later remove it.
    member this.Add (item : 'a) = 
        let itemnode = items.AddFirst item
        let mutable callbacknode = callbacks.First
        while callbacknode <> null do
            let retract = callbacknode.Value itemnode.Value
            if retract <> null then relations.AddFirst ((itemnode, callbacknode, retract)) |> ignore
            callbacknode <- callbacknode.Next
        let remove () =
            items.Remove itemnode

            // Remove and retract all relations with itemnode.
            let mutable relationnode = relations.First
            while relationnode <> null do
                let relation = relationnode.Value
                match relation with
                | (x, _, retract) when x = itemnode ->
                    let nextnode = relationnode.Next
                    relations.Remove relationnode
                    relationnode <- nextnode
                    Retract.invoke retract
                | _ -> relationnode <- relationnode.Next
        RetractAction remove

    /// Registers a callback for this feed.
    member this.Register (callback : 'a -> RetractAction) =
        let callbacknode = callbacks.AddFirst callback
        let mutable itemnode = items.First
        while itemnode <> null do
            let retract = callbacknode.Value itemnode.Value
            if retract <> null then relations.AddFirst ((itemnode, callbacknode, retract)) |> ignore
            itemnode <- itemnode.Next
        let remove () =
            callbacks.Remove callbacknode

            // Remove and retract all relations with itemnode.
            let mutable relationnode = relations.First
            while relationnode <> null do
                let relation = relationnode.Value
                match relation with
                | (_, x, retract) when x = callbacknode ->
                    let nextnode = relationnode.Next
                    relations.Remove relationnode
                    relationnode <- nextnode
                    Retract.invoke retract
                | _ -> relationnode <- relationnode.Next
        RetractAction remove

    interface CollectionFeed<'a> with
        member this.Register callback = this.Register callback

/// An event feed that applies a mapping and filter to all events from a source feed.
type MapFilterEventFeed<'a, 'b> (source : 'b event, map : 'b -> 'a option) =
    let mapCallback callback event =
        match map event with
        | Some x -> callback x
        | None -> ()

    interface EventFeed<'a> with
        member this.Register callback = source.Register (mapCallback callback)

/// A signal feed that applies a mapping to values from a source feed.
type MapSignalFeed<'a, 'b> (source : 'b signal, map : 'b -> 'a) =
    interface SignalFeed<'a> with
        member this.Current = map source.Current
        member this.Delta = 
            let mapchange (x : Change<'b>) = Some { Old = map x.Old; New = map x.New }
            match source.Delta with
            | Some sourcedelta -> Some (new MapFilterEventFeed<Change<'a>, Change<'b>> (sourcedelta, mapchange) :> Change<'a> event)
            | None -> None

/// A collection feed that applies a mapping and filter to all items from a source feed.
type MapFilterCollectionFeed<'a, 'b> (source : 'b collection, map : 'b -> 'a option) =
    interface CollectionFeed<'a> with
        member this.Register callback = source.Register (fun x ->
            match map x with
            | Some y -> callback y
            | None -> null)

/// An event feed that combines events from two source feeds.
type UnionEventFeed<'a> (sourceA : 'a event, sourceB : 'a event) =
    interface EventFeed<'a> with
        member this.Register callback = Retract.combine (sourceA.Register callback) (sourceB.Register callback)

/// A collection feed that combines items from two source feeds.
type UnionCollectionFeed<'a> (sourceA : 'a collection, sourceB : 'a collection) =
    interface CollectionFeed<'a> with
        member this.Register callback = Retract.combine (sourceA.Register callback) (sourceB.Register callback)

/// An event feed that tags events with a value from the given signal.
type TagEventFeed<'a, 'b> (event : 'a event, signal : 'b signal) =
    interface EventFeed<'a * 'b> with
        member this.Register callback = event.Register (fun x -> callback (x, signal.Current))

/// A signal feed that collates two signals into a tuple signal.
type CollateSignalFeed<'a, 'b> (sourceA : 'a signal, sourceB : 'b signal) =
    interface SignalFeed<'a * 'b> with
        member this.Current = (sourceA.Current, sourceB.Current)
        member this.Delta = new NotImplementedException () |> raise

/// An event feed that polls changes in a source feed on program updates.
type ChangePollEventFeed<'a when 'a : equality> (source : 'a signal) =
    let mutable last = source.Current
    let mutable retractUpdate : RetractAction = null
    let callbacks : Registry<'a change -> unit> = new Registry<'a change -> unit> ()

    // Polling function
    let poll time =
        let cur = source.Current
        if last <> cur then 
            let change = { Old = last; New = cur }
            for callback in callbacks do
                callback change
        last <- cur

    interface EventFeed<Change<'a>> with
        member this.Register callback =
            if callbacks.Count = 0 then retractUpdate <- Update.register poll
            let retractCallback = callbacks.Add callback
            let retract () =
                Retract.invoke retractCallback
                if callbacks.Count = 0 then Retract.invoke retractUpdate
            RetractAction retract

/// A signal feed that gives the current time in seconds.
type TimeSignalFeed private () =
    static let instance = new TimeSignalFeed ()
    let mutable time = 0.0
    let update x = time <- time + x
    do Update.register update |> ignore

    /// Gets the only instance of this type.
    static member Instance = instance

    interface SignalFeed<double> with
        member this.Current = time
        member this.Delta = None

/// Contains functions for constructing and manipulating feeds.
module Feed =
    
    /// An event feed that never fires.
    let nil<'a> = NullEventFeed<'a>.Instance :> 'a event

    /// Constructs a signal feed with a constant value.
    let constant value = new ConstSignalFeed<'a> (value) :> 'a signal

    /// A signal feed that gives the amount of real-world time, in seconds, that has
    /// elapsed since the start of the program.
    let time = TimeSignalFeed.Instance :> double signal

    /// Constructs an event feed for a native event source.
    let native (source : IEvent<'a, 'b>) =
        let control = new ControlEventFeed<'b> ()
        source.Add control.Fire
        control :> 'b event

    /// Constructs a mapped event feed.
    let mape map source = new MapFilterEventFeed<'b, 'a> (source, map >> Some) :> 'b event

    /// Constructs a mapped signal feed.
    let maps map source = new MapSignalFeed<'b, 'a> (source, map) :> 'b signal

    /// Constructs a mapped collection feed.
    let mapc map source = new MapFilterCollectionFeed<'b, 'a> (source, map >> Some) :> 'b collection

    /// Constructs a filtered event feed.
    let filtere pred source = new MapFilterEventFeed<'a, 'a> (source, fun x -> if pred x then (Some x) else None) :> 'a event

    /// Constructs a filtered collection feed.
    let filterc pred source = new MapFilterCollectionFeed<'a, 'a> (source, fun x -> if pred x then (Some x) else None) :> 'a collection

    /// Constructs a mapped and filtered event feed.
    let mapfiltere map source = new MapFilterEventFeed<'b, 'a> (source, map) :> 'b event

    /// Constructs a mapped and filtered collection feed.
    let mapfilterc map source = new MapFilterCollectionFeed<'b, 'a> (source, map) :> 'b collection

    /// Replaces the information for events from the given event feed with the given value.
    let replace value source = new MapFilterEventFeed<'a, 'b> (source, fun x -> Some value) :> 'a event

    /// Stips event information from an event feed.
    let strip source = replace () source

    /// Combines two event feeds.
    let unione a b = new UnionEventFeed<'a> (a, b) :> 'a event

    /// Combines two collection feeds.
    let unionc a b = new UnionCollectionFeed<'a> (a, b) :> 'a collection

    /// Collates two signal feeds into a tuple.
    let collate a b = new CollateSignalFeed<'a, 'b> (a, b) :> ('a * 'b) signal

    /// Tags an event feed with values from a signal feed.
    let tag a b = new TagEventFeed<'a, 'b> (a, b) :> ('a * 'b) event

    /// Gets an event feed that fires when a change occurs in the source signal feed. If it is not possible to determine
    /// exactly when a change occurs, the source will be polled on every program-wide update, and an event will be fired
    /// when a change occurs. Note that this may allow some changes to slip through, if the signal changes and returns to
    /// its original value in-between program updates.
    let delta (source : 'a signal) : 'a change event =
        match source.Delta with
        | Some sourcedelta -> mapfiltere (fun x -> if x.Old <> x.New then Some x else None) sourcedelta
        | None -> new ChangePollEventFeed<'a> (source) :> 'a change event

    /// Constructs an event feed that fires whenever the source signal changes, giving the new value.
    let change source = mape (fun x -> x.New) (delta source)

    /// Constructs an event feed that fires whenever the source signal changes from false to true.
    let rising source = mapfiltere (fun x -> if x.New = true then Some () else None) (delta source)

    /// Constructs an event feed that fires whenever the source signal changes from true to false.
    let falling source = mapfiltere (fun x -> if x.New = false then Some () else None) (delta source)

    /// Registers a callback to be called once, on the next occurence of an event in the source event feed.
    let registerOnce callback (source : 'a event) =
        let retract = ref null
        let newCallback event =
            if !retract <> null then
                retract := null
                Retract.invoke !retract
                callback event
        retract := source.Register callback